src/tree.c

```diff
@@ -425,17 +421,25 @@ enum selint_error free_policy_node(struct policy_node *to_free)
 	}
 
 	free(to_free->exceptions);
-	to_free->exceptions = NULL;
+}
+
+enum selint_error free_policy_node(struct policy_node *to_free)
+{
+	if (to_free == NULL) {
+		return SELINT_BAD_ARG;
+	}
 
-	free_policy_node(to_free->first_child);
-	to_free->first_child = NULL;
+	do {
+		struct policy_node *next = to_free->next;
 
-	free_policy_node(to_free->next);
-	to_free->next = NULL;
+		free_single_policy_node(to_free);
```

Comment:
The original order of recursing "next" first then "first_child" seems pretty weird.  I wonder if the original stack size problem would have even occurred if we'd recursed on "first_child" first.